### PR TITLE
Basic React container replicating the shape of Relay.RootContainer

### PR DESCRIPTION
**Pull request by danielmoll**
Created a basic replica of the Relay.RootContainer to manage requesting, loading and passing data.

Theory is that when we are able to use Relay we can simply update the route and rename Impart (exported name) and update the import. The Impart.RootContainer methods (excluding route) should work exactly the same as Relay.RootContainer.

In short, a basic React container that takes a React element, an async function and renders the React element with the data returned from the async function.

NOTE. 
`fetch.es6` is used for the Example to return a delayed promise to fake a data fetch.
Unlike Relay.RootContainer if none of the optional methods are supplied Relay will return `undefined` we return an empty `<div></div>`

Next steps:

Add tests (Bad Dan).
Add a Store - No need to make another request for data we already have (Nice to have).

### CODE REVIEW

**Comment by danielmoll** [comment]
@fabiosantoscode @umbobabo - care to have another look over? :)

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+Thumbs.db
+node_modules/
+_site/
+*.log
+*.log.*
+*.pid
+*.seed
+*.js
+*.html
+bundle.css
```
Comment (fabiosantoscode) [inline .gitignore]: don't forget to unignore karma.conf.js if you have one!

File: test/index.es6 [hunk 2]
```diff
@@ -0,0 +1 @@
+
```
Comment (fabiosantoscode) [inline test/index.es6]: This is waaay too short. But I'm happy if you're happy!

File: package.json [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    "doc:html": [
+      "README.md",
+      "./*.js",
+      "package.json"
+    ]
+  },
+  "config": {
+    "lint_opts": "--ignore-path .gitignore --ext .es6",
+    "testbundle_opts": "-r react -r .:./index.es6 -r ./example.js:example ./test/index.js -o testbundle.js",
+    "ghpages_files": "*.html *.css *.js assets/"
```
Comment (fabiosantoscode) [inline package.json]: You need a .gitkeep file in the assets folder.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    "doc:css": "cssnext --sourcemap example.css bundle.css",
+    "doc:css:watch": "npm run doc:css -- --watch",
+    "doc:html": "devpack-doc index standalone",
+    "doc:html:watch": "npm-watch",
+    "doc:js": "npm run prepublish && browserify -d $npm_package_config_testbundle_opts",
+    "doc:js:watch": "watchify $npm_package_config_testbundle_opts",
+    "doc:watch": "parallelshell 'npm run doc:html:watch' 'npm run doc:js:watch' 'npm run doc:css:watch'",
+    "lint": "eslint $npm_package_config_lint_opts .",
+    "pages": "git stash save -u pages-stash && git branch -D gh-pages; git checkout --orphan gh-pages && git add -f $npm_package_config_ghpages_files && git commit -anm'ghpages' && git push origin HEAD:gh-pages -f",
+    "prepages": "npm run build && npm run doc",
```
Comment (fabiosantoscode) [inline package.json]: And here's npm run build but there's no "build" script.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    "prepublish": "babel . -d . -x .es6",
```
Comment (fabiosantoscode) [inline package.json]: Here you can add `--ignore node_modules` so as to avoid recursing through the directory tree.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    "prepublish:watch": "npm run prepublish -- -w",
+    "pretest": "npm run lint",
+    "provision": "devpack-configure ./package.json",
+    "serve": "browser-sync start --server --files '*.{html,js}'",
+    "test": "karma start",
+    "watch": "parallelshell 'npm run doc:watch' 'npm run prepublish:watch' 'npm run serve'"
+  },
+  "dependencies": {
+    "react": "^0.14.0",
+    "react-dom": "^0.14.0"
```
Comment (fabiosantoscode) [inline package.json]: Depending on React is going to bring dependency hell! Keep it in devDeps and peerDeps or not at all!

**Comment by fabiosantoscode** [comment]
You might want to create a store later, or you can just use memoizee to memoize the fetching function for a quick win.

File: test/index.es6 (hunk 2 above)
Comment (keithamus) [inline test/index.es6]: Yes. Tests would be a good idea.

**Comment by keithamus** [comment]
@sebinsua `eslint-disable-line` ignores that particular line for ANY eslint rules. It disables eslint for that line.

File: package.json (hunk 6 above)
Comment (sebinsua) [inline package.json]: Is that actually our way through with this? If so I have some repos to update...

**Comment by sebinsua** [comment]
Another thing - I was wondering was what the reasons for implementing the [`Relay.RootContainer`](https://facebook.github.io/relay/docs/api-reference-relay-root-container.html) over the [`Relay.createContainer`](https://facebook.github.io/relay/docs/guides-containers.html) interface are.

I guess the former was meant to only fulfil the data requirements, while the latter was meant to describe them, but in our cases `Impart.RootContainer` attempts to both describe the data requirements and fulfil them. Obviously creating the distinction requires some complex code, but I think we should at least document the overall change in architecture and simplification it entails.

File: package.json (hunk 6 above)
Comment (fabiosantoscode) [inline package.json]: @sebinsua at least it is what we're doing at the revamp. If you guys decided to go with npm dedupe or `find node_modules | grep /react/ | xargs rm -rf; npm install react` it's fine with me too.

**Comment by sebinsua** [comment]
Looks much better to me now. Happy for a merge once it is tested.

**Comment by sebinsua** [comment]
@keithamus This is the closest to the original Relay code that we can get. Visual similarity to that is valid.

**Comment by keithamus** [comment]
@sebinsua the logic is the same either way - but we enforce code styles for a reason. I don't think we need to emulate Relay's code style, we just need to make sure we pass the parts of their test suite we care about.,

**Comment by sebinsua** [comment]
As soon as you contract the if-else braces the code will not look similar. If you want to signal diverging from the expected behaviour then that might be a good way to do so.

File: test/index.es6 (hunk 2 above)
Comment (danielmoll) [inline test/index.es6]: Added tests in latest commit with much help from @keithamus

**Comment by danielmoll** [comment]
See [StaticContainer in RelayRenderer](https://github.com/facebook/relay/blob/master/src/container/RelayRenderer.js#L279) for reference in matching Relay.RootContainer behaviour.

**Comment by sebinsua** [comment]
Looks good to me.

**Comment by umbobabo** [comment]
After talking with @danielmoll and @sebinsua it's sounds that the current stage of this implementation is ready to be released as 1.X version.
So I merge.

**Pull request merged**